[SourceKit] Report cursor info information for literals

barinsim opened the pull request:
Adds literals support for the CursorInfo request.
Previously CursorInfo just returned error for literals.
Resolves #57725

barinsim committed:
[SourceKit] Report cursor info information for literals

Adds literals support for the CursorInfo request.
Previously CursorInfo just returned error for literals.
Implements issue #57725

ahoppen reviewed (commented):
Thanks a lot for the PR, @barinsim. I’ve got a couple of minor comments inline. On a more general level, I think the problem you’ve been facing is that all of cursor info is designed to always reference a declaration, if you’re doing cursor info on a function call, we report the function declaration or if you’re doing cursor info on a variable, we return the corresponding variable declaration. 

For literals, there isn’t such a natural declaration and that’s why you needed to add a `LitExpr` to `DeclInfo`, which doesn’t really make sense since the information about a declaration (that’s what `DeclInfo` is), shouldn’t need an expression. And that’s also why you needed the `isLiteral` checks. 

What I would suggest is that we report the initializer that actually constructs eg. a `Double` from the literal (`Double.init(_builtinIntegerLiteral:)`) in cursor info. That initializer returns a value of type `Double` so you shouldn’t need all the special `isLiteral` handling anymore. You should be able to get that initializer using `getBuiltinInitializer` on `BuiltinLiteralExpr`.

Let me know if you need help with anything or have any questions about my suggestion or in general.

---

And one small thing: Could you run `git-clang-format` before committing your sources to make sure they are formatted correctly?

barinsim committed:
[SourceKit] Report init info instead of the type decl info

Addresses PR review comments.

barinsim commented on test/SourceKit/lit.local.cfg:
```diff
@@ -15,6 +15,7 @@ elif 'swift_evolve' in config.available_features:
 else:
     sk_path_sanitize = os.path.join(os.path.dirname(__file__), 'Inputs', 'sourcekitd_path_sanitize.py')
 
+    config.substitutions.append( ('%sourcekitd-test_plain', config.sourcekitd_test) )
```
I did not find a way to make a resolution of the `#colorLiteral` init work without the `-sdk %sdk` flag. Without 
the `%sourcekitd-test_plain` a temporary module cache is used where the AppKit module cache needs to be rebuilt first. This takes some time and sourcekit timeouts. Is there a standard/better way to solve this? 🤔

ahoppen replied:
I haven’t tried this myself but in complete_literal.swift we do the following, which, I think, sets the type to be used by `#colorLiteral` so we don’t need to depend on AppKit/UIKit (which we wouldn’t be able to do on Linux and Windows anyway because they don’t have AppKit/UIKit) Would you mind giving it a try? If it doesn’t work, let me know and I’ll investigate further.

https://github.com/apple/swift/blob/05919751b4c9b86f5188947d2609d32955d1b3ab/test/IDE/complete_literal.swift#L54-L58

barinsim replied:
Great that works, thank you! 👍 I removed the `sourcekitd-test_plain` along with the now unnecessary open request.

barinsim commented on test/SourceKit/CursorInfo/cursor_info.swift:
```diff
@@ -803,9 +808,52 @@ func checkAnyIsAKeyword(x: Any) {}
 
-// RUN: %sourcekitd-test -req=cursor -pos=227:14 %s -- -F %S/../Inputs/libIDE-mock-sdk -I %t.tmp %s | %FileCheck -check-prefix=CHECK94 %s
-// CHECK94: <empty cursor info; internal diagnostic: "Resolved to incomplete expression or statement.">
+// RUN: %sourcekitd-test -req=cursor -pos=23:23 %s -- -F %S/../Inputs/libIDE-mock-sdk -I %t.tmp %s | %FileCheck -check-prefix=CHECK94 %s
```
That is my oversight. I put it back with correct assertions.

barinsim committed:
[SourceKit] Fix formatting issues

barinsim commented:
@ahoppen Thank you very much for your feedback! Reporting the actual initializer definitely fits better 🙂 I hope I addressed everything.

2 quick notes:

- I preferred the `getInitializer()` and use `getBuiltinInitializer()` as a fallback, because I would argue in this case:
```swift
struct A: ExpressibleByIntegerLiteral {
    init(integerLiteral value: Int) {
        self.value = value
    }
    let value: Int
}
var a: A = 42
```
for `42` we want to report the user defined init instead of the builtin one.

- The node for the `nil` literal does not contain any initializer info. I am not sure why nor if it is expected. For now, I removed the 2 testcases testing the `nil`

ahoppen commented:
> * I preferred the `getInitializer()` and use `getBuiltinInitializer()` as a fallback, because I would argue in this case:
> 
> ```swift
> struct A: ExpressibleByIntegerLiteral {
>     init(integerLiteral value: Int) {
>         self.value = value
>     }
>     let value: Int
> }
> var a: A = 42
> ```
> 
> for `42` we want to report the user defined init instead of the builtin one.
> 

Oh, that’s even better. I didn’t think to look for `getInitializer` after I found `getBuiltinInitializer`. Good catch 👍🏽 

> * The node for the `nil` literal does not contain any initializer info. I am not sure why nor if it is expected. For now, I removed the 2 testcases testing the `nil`

Cursor info still works on `nil` if you have a custom type that conforms to `ExpressibleByNilLiteral`, e.g.

```swift
struct MyOptional: ExpressibleByNilLiteral {
  init(nilLiteral: ()) {}
}
let _: MyOptional = nil
```

The issue why we don’t get cursor info when `nil` is used for `Optional` is that the initializer doesn’t get set here

https://github.com/apple/swift/blob/05919751b4c9b86f5188947d2609d32955d1b3ab/lib/Sema/CSApply.cpp#L2628-L2634

If you wanted to, you could look into setting the initializer here as well but I’m not entirely sure what the trade-off is between performance of looking up `Optional.init(nilLiteral:)` vs the value you get out of cursor info on these `nil` literals. In either case, I think this change shouldn’t be part of this PR.

barinsim committed:
[SourceKit] Remove sourcekit-test_plain

barinsim committed:
[SourceKit] Remove unnecessary open request

ahoppen commented on test/SourceKit/CursorInfo/cursor_info.swift:
```diff
@@ -804,8 +814,47 @@ func checkAnyIsAKeyword(x: Any) {}
 // RUN: %sourcekitd-test -req=cursor -pos=227:14 %s -- -F %S/../Inputs/libIDE-mock-sdk -I %t.tmp %s | %FileCheck -check-prefix=CHECK94 %s
-// CHECK94: <empty cursor info; internal diagnostic: "Resolved to incomplete expression or statement.">
+// CHECK94: source.lang.swift.ref.function.constructor
+// CHECK94-NEXT: init(_builtinStringLiteral:utf8CodeUnitCount:isASCII:)
```
I just noticed that this shows an initializer that’s hidden from all swift interfaces (because it starts with an underscore) and takes a `Builtin` type, which is not accessible to users. What I would suggest is that instead of showing the builtin initializers, we show the type of the literal instead. See my other comment of how I think that should be done.

For other types that use an initializer that’s not a builtin initializer, I think we should still refer to the initializer.

barinsim replied:
Ok, fixed. I did the same for the object literals as their inits also seem to be undescored.

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1151,6 +1151,46 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
+        return LitExpr->getInitializer().getDecl();
+      }
+      if (auto *BuiltinLit = dyn_cast<BuiltinLiteralExpr>(LitExpr)) {
+        return BuiltinLit->getBuiltinInitializer().getDecl();
```
To get the decl of the type we are constructing, I would do the following. AFAICT, this requires to make `getInitializer` a static function instead of a lambda. And I would rename it to `getCursorInfoDeclForLiteral` + change the comment to `If `Expr` is a literal, returns the declaration that should be reported by cursor info for that initializer.`

```suggestion
      // We shouldn’t report the builtin initializer to the user because it’s underscored and not visible.
      // Instead, return the type of the literal instead.
      if (auto Ty = LitExpr->getType()) {
        if (auto NominalTy = Ty->getAs<NominalOrBoundGenericNominalType>()) {
          return NominalTy->getDecl();
        }
      }
```

barinsim replied:
I believe it still compiles fine in a lambda but I moved it into the static function since the logic became a bit more complicated 🙂 I structured this whole function a bit differently to avoid that many nested blocks.

barinsim committed:
[SourceKit] Report type decl info instead of hidden inits

barinsim commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1630,21 +1670,54 @@ static void resolveCursor(
+          switch (E->getKind()) {
+          case ExprKind::Array:
+          case ExprKind::Dictionary:
+          case ExprKind::KeyPath: {
```
It does not work and making it work is not probably trivial. Removed.

barinsim commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1630,21 +1670,54 @@ static void resolveCursor(
+                                                 getPreviousASTSnaps());
+          if (!Success) {
+            Data.Symbols.clear();
+            Data.AvailableActions.clear();
```
I was following this comment for `CursorInfoData` and clearing everything but it seems like it is not really relied upon

https://github.com/apple/swift/blob/77e694ad0d3729f4a5cab48d3cfcb3e0d34e5f74/tools/SourceKit/include/SourceKit/Core/LangSupport.h#L674-L678

I put the `pop_back` into `addCursorInfoForLiteral` as you described.

barinsim committed:
[SourceKit] Fix formatting issues

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1151,6 +1151,66 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
+    return nullptr;
+  }
+
+  bool IsObjectLiteral = dyn_cast<ObjectLiteralExpr>(E);
```
Instead of `dyn_cast` and then relying on pointer to boolean conversions, I would use `isa` here.
```suggestion
  bool IsObjectLiteral = isa<ObjectLiteralExpr>(E);
```

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1151,6 +1151,66 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
   return llvm::Error::success();
 }
 
+// If E is a literal, returns the declaration that should be reported by
```
Nitpick, you can mark `E` as a parameter.

```suggestion
// If \p E is a literal, returns the declaration that should be reported by
```

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1151,6 +1151,66 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
+    return false;
+  }
+
+  ValueDecl *InitDecl = getCursorInfoDeclForLiteral(LitExpr);
```
This no longer needs to be an initializer declaration, so I would just rename it to `D`, `InfoDecl` or something along those lines.

ahoppen reviewed (commented):
Looks very good. I just have a few minor comments.

barinsim committed:
[SourceKit] Address minor PR comments

barinsim committed:
[SourceKit] Address minor PR comments

barinsim commented:
@ahoppen Thank you very much for iterating on this with me! 🙂 I addressed the last changes you proposed. If they look alright to you, could you please run the CI?

ahoppen reviewed (approved):
Thank you @barinsim. Looks great to me

ahoppen commented:
@swift-ci Please test

ahoppen commented on lib/IDE/IDERequests.cpp:
```diff
@@ -348,8 +348,8 @@ bool CursorInfoResolver::walkToExprPost(Expr *E) {
   if (OutermostCursorExpr && isCursorOn(E, LocToResolve)) {
-    CursorInfo = new ResolvedExprStartCursorInfo(
-        CursorInfo->getSourceFile(), CursorInfo->getLoc(), OutermostCursorExpr);
+    CursorInfo = new ResolvedExprStartCursorInfo(CursorInfo->getSourceFile(),
+                                                 CursorInfo->getLoc(), E);
```
This change breaks `test/refactoring/RefactoringKind/basic.swift`:

When you have an interpolated string literal and perform cursor info on it, previously we got an interpolated string literal expression. With this change, we get the string literal expression of the first segment inside the string literal before the first interpolation and thus offer to localize the string, which we shouldn’t. I checked and it looks like this change is no longer necessary to get cursor info for literals. Could you revert it?

barinsim replied:
Yes, you are right 👍 I added a test case into the CursorInfo suite as well.

During which I found that getting info of the expression segment in the interpolated string does not work. We consider the whole interpolated string as a single token which has nested tokens within the expression segments. For example this causes that `getSourceRange()` for `ValueDecl` returns such range:

`<START>var foo = <END>"Hello \(42) world"`

This does not work if you are looking for the node of the `42`. Converting to a char based range fixed this.

ahoppen replied:
Oh, the good old char source range conversion for interpolated string literals. I so wish this wouldn’t be necessary. 🙄  Good catch though.

barinsim committed:
[SourceKit] Fix support for interpolated strings

ahoppen commented:
@swift-ci Please smoke test

fwcd commented:
Adding go-to-definition to string literals seems to have some unintended consequences in some LSP clients, e.g. when ctrl-clicking a URL in a `Package.swift` manifest in Visual Studio Code:

- https://github.com/microsoft/vscode/issues/81520

This is not Swift's or SourceKit-LSP's fault, so probably not worth working around here.

ahoppen commented:
Haha, kind of funny how it opens both the Swift interface for Swift and the URL.

But I agree that this is an editor issue and not something that needs fixing in SourceKit-LSP.
